File: parallel_programming/assignment1/CoinFlip.java
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Flips coins in parallel
 */
public class CoinFlip {
	static int numOfThreads;
	static int numOfIterations;
	static int numOfIterationsPerThread;
	static int numOfExtraIterations;
	static ArrayList<Thread> threads;
	static ArrayList<CoinFlipThread> coins;

	/**
	 * parse the command line arguments
	 * @param args command line arguments
	 */
	private static void parseArguments( String[] args ) {
		if( args.length < 2 ) {
			System.err.println( "Need 2 arguments: #threads #iterations" );
			System.exit( 0 );
		}

		// parse arguments from command line
		try {
			numOfThreads = Integer.parseInt( args[0] );
			numOfIterations = Integer.parseInt( args[1] );
		} catch( NumberFormatException e ) {
			System.err.println( "Not a proper number: " + e.getMessage() );
			System.exit( 0 );
		}
	}

	/**
	 * setup two ArrayLists.
	 */
	private static void setupArrayLists() {
		threads = new ArrayList<Thread>( numOfThreads );
		coins = new ArrayList<CoinFlipThread>( numOfThreads );
	}

	/**
	 * calculates the number of iterations per thread.
	 */
	private static void calculateIterationsPerThread() {
		numOfIterationsPerThread = numOfIterations / numOfThreads;
		if( numOfIterations % numOfThreads != 0 ) {
			numOfExtraIterations = numOfIterations - numOfThreads * numOfIterationsPerThread;
		}
	}

	/**
	 * create/setup threads.
	 */
	private static void createThreads() {
		for( int i = 0; i < numOfThreads; i++ )
		{
			// handle adding remaining extra threads to last thread
			if( numOfExtraIterations > 0 && i == numOfThreads - 1 ) {
				numOfIterationsPerThread += numOfExtraIterations;
			}
			CoinFlipThread coin = new CoinFlipThread( numOfIterationsPerThread );
			Thread thread = new Thread( coin );

			coins.add( coin );
			threads.add( thread );
		}
	}

	/**
	 * run threads
	 */
	public static void runThreads() {
		Iterator<Thread> iterator = threads.iterator();
		while( iterator.hasNext() ) {
			Thread thread = iterator.next();
			thread.start();
		}
	}

	public static void main( String[] args ) {
		int totalHeads = 0;
		long startTime = System.nanoTime(),
			 endTime = 0;

		parseArguments( args );
		setupArrayLists();
		calculateIterationsPerThread();
		long fixedStartupCostEndTime = System.nanoTime();
		long fixedStartupCost = fixedStartupCostEndTime - startTime;
		createThreads();
		long startupCostTotalThreads = System.nanoTime() - fixedStartupCostEndTime;
		System.out.println( "Fixed Startup Cost: " + fixedStartupCost );
		System.out.println( "Per Thread Startup Cost: " + startupCostTotalThreads );
		runThreads();

		// wait for threads to finish and calculate head count
		for( int i = 0; i < numOfThreads; i++ )
		{
			Thread thread = threads.get( i );
			CoinFlipThread coin = coins.get( i );
			try {
				thread.join();
			} catch( InterruptedException e )
			{
				// should not get this error
				System.err.println( "Thread got interrupted: " + e.getMessage() );
			}
			totalHeads += coin.getHeads();
		}
		endTime = System.nanoTime();

		System.out.println( totalHeads + " heads in " + numOfIterations	+ " coin tosses." );
		System.out.println( "Elapsed time: " + ( endTime - startTime ) + "ns" );
	}
}

File: parallel_programming/jruby/threads/coin_flip.rb
# Handles each CoinFlip, executes per thread
class CoinFlip
  HEADS = 0
  TAILS = 1

  attr_reader :heads_count

  def initialize( runs )
    @runs = runs
    @heads_count = 0
  end

  # runs the iterations
  def run
    1.upto( @runs ) do |i|
      # coin flip
      if rand( 2 ) == HEADS
        @heads_count += 1
      end
    end
  end
end

# calculates the iterations per thread and the extra iterations for the last thread
def calculate_iterations( num_of_iterations, num_of_threads )
  iterations_per_thread = num_of_iterations / num_of_threads
  extra_iterations =
    if( num_of_iterations % num_of_threads != 0 )
      num_of_iterations - num_of_threads * iterations_per_thread;
    else
      0
    end

  [iterations_per_thread, extra_iterations]
end

if $0 == __FILE__
  if ARGV.length < 2
    $stderr.puts "need to specify number of iterations and number of threads."
    exit
  end

  threads = Array.new
  coins = Array.new
  num_of_iterations = ARGV.shift.to_i
  num_of_threads = ARGV.shift.to_i
  iterations_per_thread, extra_iterations = calculate_iterations( num_of_iterations, num_of_threads )
  start_time = Time.now

  1.upto( num_of_threads ) do |thread_num|
    iterations = iterations_per_thread
    iterations += extra_iterations if thread_num == num_of_iterations
    threads << Thread.new( iterations ) do |runs|
      coin_flip = CoinFlip.new( runs )
      coins << coin_flip
      coin_flip.run
    end
  end

  threads.each {|thread| thread.join }
  total_heads = coins.inject( 0 ) {|sum, coin| coin.heads_count + sum }

  end_time = Time.now

  puts "#{total_heads} heads in #{num_of_iterations} coin tosses."
  puts "Start time: #{start_time}"
  puts "End time: #{end_time}"
  puts "Elapsed time: #{end_time - start_time} s"
end
